### ISSUE TITLE
Prototype for type inference IR pass

### ISSUE THREAD

**Issue by radeusgd** [enso-org/enso#8590]
The first iteration of the type inference IR pass.

- [x] Add a new IR pass and associated metadata for each expression.
  - [x] The IR pass should be enabled by a feature flag; disabled by default.
    - This is because we want to be able to measure how it impacts performance, once #8419 is integrated.
- [x] Initialize it from the existing type ascriptions.
- [x] Traverse the expression tree and try to thread through the type information
  - [x] If the ascription is at the output (outer) expression of the function, we can record the return type of the function.
  - [x] See how to record the arity of the functions.
  - ~~Special logic for `case of` and `if` (as they need to be polymorphic to do any good, but the first iteration of the type system is not polymorphic).~~
  - For now, the type inference will rely on the 'functional' subset - i.e. applying to a function with checked arguments, or getting back a value from a function whose return type is known, yields known types.
  - In this iteration, no logic for `o.member_method`s, because that requires knowing the methods defined on a Type.
    - This is a larger endeavour and will be done later. We will need to extend `BindingsMap.ResolvedType` to know not only the constructors, but also methods on a type. Separate ticket to be created later.
- Displaying the inferred type info on the tree for debug purposes.
- Possibly reporting 'obvious' type errors if detected.
- See what is needed to report the types to the IDE.
  - The intention is to get compile-time type info for earlier loading of widgets.
  - This won't get us much without the `o.member_method`s - because 99% of widgets are on those, so it may make sense to do it later with the next iteration.

**Comment by radeusgd** [comment]
Stuff to do before first PR:
- [x] fix spurious Not_Invokable warnings in constructr signatures like `Vector Integer`,
- [ ] fix persistance serialization errors,
- [x] refactor: cleanup Type Inference logic, better manage common utils (package repository access, context)
- [x] disable object type checks until conversions are resolved,
- ~~check if I can infer type of `self`~~ (delayed after #8805 is implemented),
- [x] fix logging,
- [x] cleanup (move TypeCompatibility outside etc.)
- [x] refactor: try avoiding duplicating the logic of `IrToTruffle::processName` within the inference - we should try to create some abstraction so that the logic is kept in just one place and accessible from both contexts
- [x] ensure we can run `Base_Tests` with type check flag,
- [x] test consistency: if a `Not_Invokable` warning is raised, check that it will actually happen at runtime.

**Comment by radeusgd** [comment]
Currently `from Standard.Base import all` fails with
```
Execution finished with an error: java.lang.IllegalStateException: Internal error: type signature contained _already resolved_ reference to type Standard.Base.Data.Json.Invalid_JSON, but now that type cannot be found in the bindings map.
```

- [x] Investigate and try finding a smaller repro.
